ml-prims/src/random/permute.h

```diff
@@ -16,135 +16,99 @@
 
 #pragma once
 
-#include <cub/cub.cuh>
+#include <cooperative_groups.h>
+#include <memory>
 #include "cuda_utils.h"
-#include "decoupled_lookback.h"
 #include "vectorized.h"
 
-
 namespace MLCommon {
 namespace Random {
 
-template <typename IntType, int NumItems>
-DI void getIndices(IntType (&idx)[NumItems], IntType a, IntType b, int k,
-                   IntType n) {
-    IntType tid = NumItems * (threadIdx.x + (blockIdx.x * blockDim.x));
-    IntType mask = (1 << k) - 1;
-#pragma unroll
-    for (int i = 0; i < NumItems; ++i)
-        idx[i] = ((a * (tid + i)) + b) & mask;
-}
+template <typename Type, typename IntType, int TPB, bool rowMajor>
+__global__ void permuteKernel(IntType* perms, Type* out, const Type* in,
+                              IntType a, IntType b, IntType N, IntType D) {
+    namespace cg = cooperative_groups;
+    const int WARP_SIZE = 32;
 
-template <typename IntType, int NumItems, int TPB>
-DI void compaction(const IntType (&idx)[NumItems], IntType (&outIdx)[NumItems],
-                   DecoupledLookBack<int>& dlb, IntType N) {
-#pragma unroll
-    for (int i = 0; i < NumItems; ++i)
-        outIdx[i] = idx[i] < N;
-    typedef cub::BlockScan<IntType, TPB> BlockScan;
-    __shared__ typename BlockScan::TempStorage temp_storage;
-    BlockScan(temp_storage).ExclusiveSum(outIdx, outIdx);
-    IntType thisBlockSum = 0;
-    if(threadIdx.x == blockDim.x - 1)
-        thisBlockSum = outIdx[NumItems-1] + (idx[NumItems - 1] < N);
-    auto prefix = dlb(thisBlockSum);
-#pragma unroll
-    for (int i = 0; i < NumItems; ++i)
-        outIdx[i] += prefix;
-}
+    int tid = threadIdx.x + blockIdx.x*blockDim.x;
 
-///@todo: fix shared mem bank conflicts
-template <typename IntType, int NumItems, int TPB>
-DI void writePermIndices(IntType* perms, const IntType (&idx)[NumItems],
-                         const IntType (&outIdx)[NumItems], IntType* s_perms,
-                         IntType N, int k) {
-    int tid = threadIdx.x;
-#pragma unroll
-    for (int i = 0; i < NumItems; ++i) {
-        int pos = tid * NumItems + i;
-        s_perms[pos] = outIdx[i];
-        s_perms[pos + NumItems * TPB] = idx[i];
+    // having shuffled input indices and coalesced output indices appears
+    // to be preferrable to the reverse, especially for column major
+    IntType inIdx = ((a * int64_t(tid)) + b) % N;
```

Comment:
I have to try if this is still correct. The reason why i have the explicit type there is that this fails badly if the multiplication overflows (both with unsigned and signed types). If someone were to try and save space on the permutation output array by making IntType short for example this could easily happen.

Reply:
Using IntType for all the dimensions as well as the type of the permutation array is maybe a bit questionable anyway? Should the size/dimension parameters maybe be of a fixed type? size_t? Or at least of their own template type. The above issue would still apply though in case someone accidentally passed N as a very small type.

Reply:
good point! Better to expose an `IdxType` template param for dimensions as well as for indexing logic.